### ISSUE TITLE
head for segmentation model

### ISSUE THREAD

**Issue by kiucho** [itsprakhar/Simple-DinoV2-Classification#3]
hello. First of all, thank you for sharing your implementation.

I am fine-tune DinoV2 on semantic segmentation task.
While looking at `model.py`, I figured out that classification model gets linear head but segmentation model gets conv head.
Is there any reason why segmentation model gets conv head?

Thank you.

**Comment by itsprakhar** [comment]
Hello @kiucho, This is just a design choice,  I have used conv layers to transform the (embedding_size, h, w) tensor to (num_classes, h, w) tensor While using some upsampling layers to increase the size of the segmentation map. The 3x3 convs can be replaced with 1x1 convs (conceptually similar to having linear layer without biases).

Hope this addresses your question, feel free to ask any more questions or provide feedback.

**Comment by kiucho** [comment]
Thank you for your sincere reply! I have one more question for you. 
I'm new to the vision field, but is there a reason why you specified a separate transform for image and mask?
```
img_transform = transforms.Compose([
    transforms.Resize((14*32,14*32)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

mask_transform = transforms.Compose([
    transforms.Resize((32*4,32*4)),
    transforms.ToTensor(),
])
```
resized them to different sizes, and normalized only the image?

Thank you.

**Comment by fefespn** [comment]
regarding this issue, can we use can you give example how can we use huggingface mas2former or Oneformer as the head (in place of the conv and linear)?

**Comment by renxiaosa00** [comment]
> Thank you for your sincere reply! I have one more question for you. I'm new to the vision field, but is there a reason why you specified a separate transform for image and mask?
> 
> ```
> img_transform = transforms.Compose([
>     transforms.Resize((14*32,14*32)),
>     transforms.ToTensor(),
>     transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
> ])
> 
> mask_transform = transforms.Compose([
>     transforms.Resize((32*4,32*4)),
>     transforms.ToTensor(),
> ])
> ```
> 
> resized them to different sizes, and normalized only the image?
> 
> Thank you.

I have same question.

**Comment by imagra93** [comment]
> > Thank you for your sincere reply! I have one more question for you. I'm new to the vision field, but is there a reason why you specified a separate transform for image and mask?
> > ```
> > img_transform = transforms.Compose([
> >     transforms.Resize((14x32,14x32)),
> >     transforms.ToTensor(),
> >     transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
> > ])
> > 
> > mask_transform = transforms.Compose([
> >     transforms.Resize((32x4,32x4)),
> >     transforms.ToTensor(),
> > ])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > resized them to different sizes, and normalized only the image?
> > Thank you.
> 
> I have same question.

On one hand, **img_transform** does 3 things: 

- resizes the input images to (14x32,14x32) because the backbone dinov2 expects images of that size.
- converts input arrays into torch tensors.
- normalizes input images to the same mean and std as the images dinov2 was trained on.

On the other hand, **mask_transform**:

- converts them into tensors as well, but does not normalize them. Each value/pixel of the mask represents 1 class (out of as many as your segmentation problem has, i.e. `num_classes`) and we want these classes to remain unchanged (values between 0 and `num_classes`-1).
- Dinov2 expects input images of size (14x32,14x32), but the convolutional head proposed in this repository outputs a tensor of size (32x4,32x4) that intends to represent/learn the segmentation mask but in a lower resolution. Therefore, in order to obtain the loss (or metrics) pixel by pixel (compare the output of the model with your truth labels), we need our original masks resized accordingly.

**Comment by fyang93** [comment]
> > > Thank you for your sincere reply! I have one more question for you. I'm new to the vision field, but is there a reason why you specified a separate transform for image and mask?
> > > ```
> > > img_transform = transforms.Compose([
> > >     transforms.Resize((14x32,14x32)),
> > >     transforms.ToTensor(),
> > >     transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
> > > ])
> > > 
> > > mask_transform = transforms.Compose([
> > >     transforms.Resize((32x4,32x4)),
> > >     transforms.ToTensor(),
> > > ])
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > resized them to different sizes, and normalized only the image?
> > > Thank you.
> > 
> > 
> > I have same question.
> 
> On one hand, **img_transform** does 3 things:
> 
> * resizes the input images to (14x32,14x32) because the backbone dinov2 expects images of that size.
> * converts input arrays into torch tensors.
> * normalizes input images to the same mean and std as the images dinov2 was trained on.
> 
> On the other hand, **mask_transform**:
> 
> * converts them into tensors as well, but does not normalize them. Each value/pixel of the mask represents 1 class (out of as many as your segmentation problem has, i.e. `num_classes`) and we want these classes to remain unchanged (values between 0 and `num_classes`-1).
> * Dinov2 expects input images of size (14x32,14x32), but the convolutional head proposed in this repository outputs a tensor of size (32x4,32x4) that intends to represent/learn the segmentation mask but in a lower resolution. Therefore, in order to obtain the loss (or metrics) pixel by pixel (compare the output of the model with your truth labels), we need our original masks resized accordingly.

AFAIK, DINOv2 uses 518*518 (14x37,14x37) as the input size, instead of (14x32,14x32)

**Comment by odusseys** [comment]
Hi - you should consider adding batch normalization after your convolutions and removing the bias term

I have also played with some variations, especially using inverse convolutions (such as 2x2 transposed conv with a stride of 2) to make the model able to learn smarter upsamplings. Just upsampling smoothes out the patch features which means the resulting upsampled patch really doesn't carry much more information. This in particular gives more crisp mask boundaries, and allows to segment finer details ; whereas upsampling leads to broad  smooth boundaries.
